src/HashSource.ts

```diff
@@ -0,0 +1,84 @@
+import Evented from 'dojo-core/Evented';
+import global from 'dojo-core/global';
+import { Handle } from 'dojo-core/interfaces';
+import { createHandle } from 'dojo-core/lang';
+import has from 'dojo-core/has';
+import { NavigationArgs, RouterSource } from './routing';
+
+export default class HashSource extends Evented implements RouterSource {
+	protected _current: { hash: string; path: string };
+	protected _handle: Handle;
+	protected _prefix: string;
+	protected _prefixPattern: RegExp;
+
+	get currentPath(): string {
+		return global.location.hash.replace('#', '');
+	}
+
+	get prefix(): string {
+		return this._prefix;
+	}
+	set prefix(prefix: string) {
+		this._prefix = prefix || '';
+		this._prefixPattern = new RegExp('^#?' + this._prefix, 'g');
+	}
+
+	constructor(prefix: string = '!') {
+		if (!has('host-browser')) {
+			throw new Error('HashSource requires a browser.');
```

Comment:
The `host-browser` test already checks for `global.location`, but I guess we don't technically have to care whether `global.document` exists as well.